libsplinter/src/network/peer_manager/peer_map.rs

```diff
@@ -0,0 +1,162 @@
+    /// the peer.
+    pub fn remove(&mut self, peer_id: &str) -> Option<String> {
+        debug!("Removing peer: {}", peer_id);
+        self.redirects
+            .retain(|_, target_peer_id| target_peer_id != peer_id);
+        if let Some(peer_metadata) = self.peers.remove(&peer_id.to_string()) {
+            for endpoint in peer_metadata.endpoints.iter() {
+                self.endpoints.remove(endpoint);
+            }
+
+            Some(peer_metadata.active_endpoint)
+        } else {
+            None
+        }
+    }
+
+    /// Updates a peer id, and creates a redirect for the old id to the given new one.
+    ///
+    /// Additionally, it updates all of the old redirects to point to the given new one.
+    pub fn update_peer_id(
+        &mut self,
+        old_peer_id: String,
+        new_peer_id: String,
+    ) -> Result<(), PeerUpdateError> {
+        if let Some(peer_metadata) = self.peers.remove(&old_peer_id) {
+            let mut new_peer_metadata = peer_metadata.clone();
+            for endpoint in peer_metadata.endpoints.iter() {
+                self.endpoints
+                    .insert(endpoint.to_string(), new_peer_id.clone());
+            }
+
+            new_peer_metadata.id = new_peer_id.clone();
+            self.peers.insert(new_peer_id.clone(), new_peer_metadata);
+
+            // update the old forwards
+            for (_, v) in self
+                .redirects
+                .iter_mut()
+                .filter(|(_, v)| **v == old_peer_id)
+            {
+                *v = new_peer_id.clone()
+            }
+
+            self.redirects.insert(old_peer_id, new_peer_id);
+
+            Ok(())
+        } else {
+            Err(PeerUpdateError(format!(
+                "Unable to update {} to {}",
+                old_peer_id, new_peer_id
+            )))
+        }
+    }
+
+    /// Updates the active endpoint for a peer
+    pub fn update_active_endpoint(
+        &mut self,
+        peer_id: &str,
+        active_endpoint: String,
+    ) -> Result<(), PeerUpdateError> {
+        if let Some(peer_metadata) = self.peers.get(peer_id) {
+            let mut new_metadata = peer_metadata.clone();
+            new_metadata.active_endpoint = active_endpoint;
+            self.peers.insert(peer_id.to_string(), new_metadata);
+            Ok(())
+        } else {
+            Err(PeerUpdateError(format!(
+                "Unable to to update active endpoint for {}",
+                peer_id
+            )))
+        }
+    }
+
+    /// Returns the endpoint for the given peer id
+    pub fn get_peer_from_endpoint(&self, endpoint: &str) -> Option<String> {
+        self.endpoints.get(endpoint).cloned()
+    }
+
+    pub fn set_disconnected(&mut self, peer_id: &str, attempts: u64) {
+        self.disconnected.insert(peer_id.to_string(), attempts);
+    }
+
+    pub fn set_connected(&mut self, peer_id: &str) {
+        self.disconnected.remove(peer_id);
+    }
+}
```

Comment:
Probably want some tests for this module

Reply:
added